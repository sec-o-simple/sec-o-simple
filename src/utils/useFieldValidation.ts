import useValidationStore from '@/utils/useValidationStore'

export function useFieldValidation(path?: string) {
  const validationStore = useValidationStore()

  const messages = !path ? [] : validationStore.getMessagesForPath(path)
  const errorMessages = messages.filter((m) => m.severity === 'error')
  const warningMessages = messages.filter((m) => m.severity === 'warning')
  const infoMessages = messages.filter((m) => m.severity === 'info')

  return {
    messages,
    hasErrors: errorMessages.length > 0,
    hasWarnings: warningMessages.length > 0,
    hasInfos: infoMessages.length > 0,
    errorMessages,
    warningMessages,
    infoMessages,
    isTouched: !path ? true : validationStore.isFieldTouched(path),
    markFieldAsTouched: validationStore.markFieldAsTouched,
  }
}
