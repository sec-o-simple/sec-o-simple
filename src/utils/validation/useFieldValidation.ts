import useValidationStore from '@/utils/validation/useValidationStore'

export function useFieldValidation(path?: string) {
  const validationStore = useValidationStore()

  const messages = !path ? [] : validationStore.getMessagesForPath(path)
  const errorMessages = messages.filter((m) => m.severity === 'error')
  const warningMessages = messages.filter((m) => m.severity === 'warning')
  const infoMessages = messages.filter((m) => m.severity === 'info')

  return {
    messages,
    hasErrors: errorMessages.length > 0,
    hasWarnings: warningMessages.length > 0,
    hasInfos: infoMessages.length > 0,
    errorMessages,
    warningMessages,
    infoMessages,
    // For now we mark all fields as touched until we have a better way to handle this
    isTouched: true,
    markFieldAsTouched: validationStore.markFieldAsTouched,
  }
}
