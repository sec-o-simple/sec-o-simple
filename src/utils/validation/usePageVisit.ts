import { useEffect, useState } from 'react'
import { useLocation } from 'react-router'
import useValidationStore from './useValidationStore'

/**
 * A hook that tracks page visits and manages page visit state.
 * It marks a page as visited after a brief delay and tracks whether the current page
 * has been previously visited.
 *
 * @returns {boolean} Returns true if the current page has been previously visited, false otherwise
 *
 * @example
 * function MyComponent() {
 *   const hasVisited = usePageVisit();
 *   return hasVisited ? <div>Welcome back!</div> : <div>First time here!</div>;
 * }
 */
export default function usePageVisit() {
  const visitPage = useValidationStore((state) => state.visitPage)
  const hasVisitedPage = useValidationStore((state) => state.hasVisitedPage)
  const [visitedPage, setHasVisitedPage] = useState(false)
  const location = useLocation()

  useEffect(() => {
    // Handle null location case by throwing an error
    if (!location) {
      throw new TypeError("Cannot read properties of null (reading 'pathname')")
    }

    // Check if page was previously visited
    if (hasVisitedPage(location.pathname)) {
      setHasVisitedPage(true)
    }

    // Mark page as visited after short delay
    const timeoutId = setTimeout(() => {
      visitPage(location.pathname)
    }, 500)

    // Cleanup timeout if user leaves page before delay
    return () => clearTimeout(timeoutId)
  }, [location, visitPage, hasVisitedPage])

  return visitedPage
}
