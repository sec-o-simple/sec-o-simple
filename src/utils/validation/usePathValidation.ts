import useValidationStore from './useValidationStore'

type HasErrorFunction = (errorPaths: string[]) => boolean

const validationSections: Record<string, HasErrorFunction> = {
  '/document-information/general': (errorPaths) => {
    return ['/document/title', '/document/tracking/id', '/document/lang'].some(
      (path) => errorPaths.includes(path),
    )
  },
  '/tracking': (errorPaths) => {
    return (
      errorPaths.includes('/document/tracking/status') ||
      errorPaths.some((path) =>
        path.startsWith('/document/tracking/revision_history'),
      )
    )
  },
  '/document-information/notes': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/document/notes'))
  },
  '/document-information/publisher': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/document/publisher'))
  },
  '/document-information/acknowledgments': (errorPaths) => {
    return errorPaths.some((path) =>
      path.startsWith('/document/acknowledgments'),
    )
  },
  '/document-information/references': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/document/references'))
  },
  '/products': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/products'))
  },
  '/vulnerabilities': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/vulnerabilities'))
  },
}

export function usePathValidation(path: string) {
  const messages = useValidationStore((state) => state.messages)
  const errorPaths = messages
    .filter((m) => m.severity === 'error')
    .map((e) => e.path)
  // const hasVisitedPage = useValidationStore((state) => state.hasVisitedPage)
  if (validationSections[path]) {
    const hasErrors = validationSections[path](errorPaths)
    return {
      hasErrors,
      // For now we mark all fields as touched until we have a better way to handle this
      // hasVisited: hasVisitedPage(path),
      hasVisited: true,
    }
  }

  return {
    hasErrors: false,
    // For now we mark all fields as touched until we have a better way to handle this
    // hasVisited: hasVisitedPage(path),
    hasVisited: true,
  }
}
