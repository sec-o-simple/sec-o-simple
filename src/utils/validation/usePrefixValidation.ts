import useValidationStore from "./useValidationStore"

export function usePrefixValidation(prefix: string) {
    const messages = useValidationStore((state) => state.messages)
    const errorPaths = messages
        .filter((m) => m.severity === 'error')
        .map((e) => e.path)

    const hasErrors = errorPaths.some((path) => path.startsWith(prefix + '/'))

    return {
        hasErrors,
    }
}