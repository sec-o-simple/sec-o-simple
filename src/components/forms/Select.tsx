import { useFieldValidation } from '@/utils/useFieldValidation'
import { Select as HeroUISelect, SelectProps } from '@heroui/select'

export default function Select(
  props: SelectProps & { csafPath?: string; isTouched?: boolean },
) {
  const {
    placeholder,
    labelPlacement,
    variant,
    csafPath,
    isTouched,
    onChange,
    ...rest
  } = props
  const validation = useFieldValidation(csafPath)

  const handleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (csafPath) {
      // Only after touching the field, we want to show errors.
      validation.markFieldAsTouched(csafPath)
    }
    onChange?.(e)
  }

  return (
    <HeroUISelect
      labelPlacement={labelPlacement ?? 'outside'}
      placeholder={placeholder ?? ' '}
      variant={variant ?? 'bordered'}
      classNames={{
        trigger: 'border-1 shadow-none',
      }}
      onChange={handleChange}
      errorMessage={validation.errorMessages.map((m) => m.message).join(', ')}
      isInvalid={validation.hasErrors && (isTouched || validation.isTouched)}
      {...rest}
    />
  )
}
