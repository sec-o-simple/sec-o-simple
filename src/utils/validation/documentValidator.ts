import { TDocumentStore } from '@/utils/useDocumentStore'
import { createCSAFDocument } from '../csafExport/csafExport'

import * as basic from '@secvisogram/csaf-validator-lib/basic.js'
import validate from '@secvisogram/csaf-validator-lib/validate.js'
import { TConfig } from '../useConfigStore'
import { ValidationMessage } from './useValidationStore'

const tests: unknown[] = [...Object.values(basic)]

export interface ValidationResult {
  isValid: boolean
  messages: ValidationMessage[]
}

export async function validateDocument(
  documentStore: TDocumentStore,
  getRelationshipFullProductName: (
    sourceVersionId: string,
    targetVersionId: string,
    category: string,
  ) => string,
  config?: TConfig,
): Promise<ValidationResult> {
  try {
    const csafDocument = createCSAFDocument(
      documentStore,
      getRelationshipFullProductName,
      config,
    )
    const result = await validate(tests, csafDocument)

    // Collect all messages with their severity
    let messages: ValidationMessage[] = []

    result.tests.forEach((test) => {
      test.errors.forEach((error) => {
        messages.push({
          path: error.instancePath,
          message: error.message,
          severity: 'error',
        })
      })

      test.warnings.forEach((warning) => {
        messages.push({
          path: warning.instancePath,
          message: warning.message,
          severity: 'warning',
        })
      })

      test.infos.forEach((info) => {
        messages.push({
          path: info.instancePath,
          message: info.message,
          severity: 'info',
        })
      })
    })

    return {
      // We want to use result.isValid when we implemented all fields
      // but for now we use the messages length to determine if the document is valid
      isValid: messages.filter((m) => m.severity === 'error').length === 0,
      messages,
    }
  } catch (error) {
    return {
      isValid: false,
      messages: [],
    }
  }
}
