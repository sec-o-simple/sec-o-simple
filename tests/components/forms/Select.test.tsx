import { render, screen, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import Select from '../../../src/components/forms/Select'
import { useFieldValidation } from '../../../src/utils/validation/useFieldValidation'

// Mock the validation hook
vi.mock('../../../src/utils/validation/useFieldValidation', () => ({
  useFieldValidation: vi.fn(() => ({
    errorMessages: [],
    hasErrors: false,
    isTouched: false,
    markFieldAsTouched: vi.fn(),
  })),
}))

// Mock HeroUI Select component
vi.mock('@heroui/select', () => ({
  Select: vi.fn(({ children, errorMessage, isInvalid, classNames, ...props }: any) => (
    <select 
      {...props} 
      data-testid="hero-select"
      data-error-message={errorMessage || ''}
      data-is-invalid={isInvalid?.toString() || 'false'}
      data-classnames={typeof classNames === 'object' ? JSON.stringify(classNames) : classNames}
    >
      {children}
    </select>
  )),
}))

const mockUseFieldValidation = vi.mocked(useFieldValidation)

describe('Select', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    mockUseFieldValidation.mockReturnValue({
      messages: [],
      errorMessages: [],
      warningMessages: [],
      infoMessages: [],
      hasErrors: false,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    })
  })

  it('should render with default props', () => {
    const { container } = render(<Select><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    expect(select).toBeInTheDocument()
    expect(select).toHaveAttribute('variant', 'bordered')
    expect(select).toHaveAttribute('labelPlacement', 'outside')
    expect(select).toHaveAttribute('placeholder', ' ')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should accept custom props', () => {
    render(
      <Select
        label="Custom Label"
        placeholder="Custom placeholder"
        variant="underlined"
        labelPlacement="inside"
      >
        <option>Test</option>
      </Select>
    )
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('label', 'Custom Label')
    expect(select).toHaveAttribute('placeholder', 'Custom placeholder')
    expect(select).toHaveAttribute('variant', 'underlined')
    expect(select).toHaveAttribute('labelPlacement', 'inside')
  })

  it('should apply custom classNames for trigger', () => {
    render(<Select><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('data-classnames', JSON.stringify({
      trigger: 'border-1 shadow-none',
    }))
  })

  it('should handle onChange events', async () => {
    const user = userEvent.setup()
    const onChange = vi.fn()
    
    render(<Select onChange={onChange}><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    fireEvent.change(select, { target: { value: 'option1' } })
    
    expect(onChange).toHaveBeenCalled()
  })

  it('should mark field as touched on change with csafPath', () => {
    const mockMarkFieldAsTouched = vi.fn()
    const mockValidation = {
      messages: [],
      errorMessages: [],
      warningMessages: [],
      infoMessages: [],
      hasErrors: false,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: mockMarkFieldAsTouched,
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<Select csafPath="test.field"><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    fireEvent.change(select, { target: { value: 'option1' } })
    
    expect(mockMarkFieldAsTouched).toHaveBeenCalledWith('test.field')
  })

  it('should display validation errors when field has errors and is touched', () => {
    const mockValidation = {
      messages: [{ message: 'Field is required', severity: 'error' }],
      errorMessages: [{ message: 'Field is required' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: true,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<Select csafPath="test.field"><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('data-error-message', 'Field is required')
    expect(select).toHaveAttribute('data-is-invalid', 'true')
  })

  it('should display validation errors when field has errors and isTouched prop is true', () => {
    const mockValidation = {
      messages: [{ message: 'Invalid selection', severity: 'error' }],
      errorMessages: [{ message: 'Invalid selection' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<Select csafPath="test.field" isTouched><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('data-error-message', 'Invalid selection')
    expect(select).toHaveAttribute('data-is-invalid', 'true')
  })

  it('should show error message but not mark as invalid when field has errors but is not touched', () => {
    const mockValidation = {
      messages: [{ message: 'Field is required', severity: 'error' }],
      errorMessages: [{ message: 'Field is required' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<Select csafPath="test.field"><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('data-error-message', 'Field is required')
    expect(select).toHaveAttribute('data-is-invalid', 'false')
  })

  it('should handle multiple validation errors', () => {
    const mockValidation = {
      messages: [
        { message: 'Field is required', severity: 'error' },
        { message: 'Invalid format', severity: 'error' }
      ],
      errorMessages: [
        { message: 'Field is required' },
        { message: 'Invalid format' }
      ],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: true,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<Select csafPath="test.field"><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('data-error-message', 'Field is required, Invalid format')
  })

  it('should work without csafPath (no validation)', () => {
    const onChange = vi.fn()
    
    render(<Select onChange={onChange}><option>Test</option></Select>)
    
    const select = screen.getByTestId('hero-select')
    fireEvent.change(select, { target: { value: 'option1' } })
    
    expect(onChange).toHaveBeenCalled()
    expect(select).toHaveAttribute('data-error-message', '')
    expect(select).toHaveAttribute('data-is-invalid', 'false')
  })

  it('should pass through other props to HeroUI Select', () => {
    render(
      <Select
        disabled
        required
        name="test-select"
        data-testid="custom-select"
      >
        <option>Test</option>
      </Select>
    )
    
    const select = screen.getByTestId('hero-select')
    expect(select).toHaveAttribute('disabled', '')
    expect(select).toHaveAttribute('required', '')
    expect(select).toHaveAttribute('name', 'test-select')
  })

  describe('Snapshots', () => {
    it('should render default state', () => {
      const { container } = render(<Select><option>Test</option></Select>)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should render with error state', () => {
      const mockValidation = {
        messages: [{ message: 'Required field', severity: 'error' }],
        errorMessages: [{ message: 'Required field' }],
        warningMessages: [],
        infoMessages: [],
        hasErrors: true,
        hasWarnings: false,
        hasInfos: false,
        isTouched: true,
        markFieldAsTouched: vi.fn(),
      }
      
      mockUseFieldValidation.mockReturnValue(mockValidation)
      
      const { container } = render(<Select csafPath="test.field"><option>Test</option></Select>)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should render with custom variant', () => {
      const { container } = render(<Select variant="flat"><option>Test</option></Select>)
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
