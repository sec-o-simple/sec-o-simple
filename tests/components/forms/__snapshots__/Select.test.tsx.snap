// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > Snapshots > should render default state 1`] = `
<select
  data-classnames="{"trigger":"border-1 shadow-none"}"
  data-error-message=""
  data-is-invalid="false"
  data-testid="hero-select"
  labelplacement="outside"
  placeholder=" "
  variant="bordered"
>
  <option>
    Test
  </option>
</select>
`;

exports[`Select > Snapshots > should render with custom variant 1`] = `
<select
  data-classnames="{"trigger":"border-1 shadow-none"}"
  data-error-message=""
  data-is-invalid="false"
  data-testid="hero-select"
  labelplacement="outside"
  placeholder=" "
  variant="flat"
>
  <option>
    Test
  </option>
</select>
`;

exports[`Select > Snapshots > should render with error state 1`] = `
<select
  data-classnames="{"trigger":"border-1 shadow-none"}"
  data-error-message="Required field"
  data-is-invalid="true"
  data-testid="hero-select"
  labelplacement="outside"
  placeholder=" "
  variant="bordered"
>
  <option>
    Test
  </option>
</select>
`;

exports[`Select > should render with default props 1`] = `
<select
  data-classnames="{"trigger":"border-1 shadow-none"}"
  data-error-message=""
  data-is-invalid="false"
  data-testid="hero-select"
  labelplacement="outside"
  placeholder=" "
  variant="bordered"
>
  <option>
    Test
  </option>
</select>
`;
